src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs

```diff
@@ -51,142 +34,185 @@ private ExpressionSyntax ParseInterpolatedStringToken()
+                    }
+
+                    // Add a token for text following the last interpolation
+                    var lastText = originalText[new Range(interpolations[^1].CloseBraceRange.End, closeQuoteRange.Start)];
+                    if (lastText.Length > 0)
+                    {
+                        var token = MakeStringToken(lastText, lastText, isVerbatim, SyntaxKind.InterpolatedStringTextToken);
+                        builder.Add(SyntaxFactory.InterpolatedStringText(token));
+                    }
                 }
+
+                CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> result = builder;
+                _pool.Free(builder);
+                return result;
             }
 
-            interpolations.Free();
-            var result = SyntaxFactory.InterpolatedStringExpression(openQuote, builder, closeQuote);
-            _pool.Free(builder);
-            if (error != null)
+            SyntaxToken getCloseQuote()
             {
-                result = result.WithDiagnosticsGreen(new[] { error });
+                // Make a token for the close quote " (even if it was missing)
+                var closeQuoteText = originalText[closeQuoteRange];
+                return closeQuoteText == ""
+                    ? SyntaxFactory.MissingToken(leading: null, SyntaxKind.InterpolatedStringEndToken, trailing: originalToken.GetTrailingTrivia())
+                    : SyntaxFactory.Token(leading: null, SyntaxKind.InterpolatedStringEndToken, closeQuoteText, closeQuoteText, originalToken.GetTrailingTrivia());
             }
+        }
 
-            Debug.Assert(originalToken.ToFullString() == result.ToFullString()); // yield from text equals yield from node
-            return CheckFeatureAvailability(result, MessageID.IDS_FeatureInterpolatedStrings);
+        private static InterpolationSyntax ParseInterpolation(CSharpParseOptions options, string text, Lexer.Interpolation interpolation, bool isVerbatim)
+        {
+            // Grab the text from after the { all the way to the start of the } (or the start of the : if present). This
+            // will be used to parse out the expression of the interpolation.
+            //
+            // The parsing of the open brace, close brace and colon is specially handled in ParseInterpolation below.
+            var expressionText = text[new Range(
+                interpolation.OpenBraceRange.End,
+                interpolation.HasColon ? interpolation.ColonRange.Start : interpolation.CloseBraceRange.Start)];
+
+            using var tempLexer = new Lexer(SourceText.From(expressionText), options, allowPreprocessorDirectives: false, interpolationFollowedByColon: interpolation.HasColon);
+
+            // First grab any trivia right after the {, it will be trailing trivia for the { token.
+            var openTokenTrailingTrivia = tempLexer.LexSyntaxTrailingTrivia().Node;
+            var openTokenText = text[interpolation.OpenBraceRange];
```

Comment:
this is the thrust of the change.  the initial trivia is actually trailing trivia on the preceding token.

Reply:
What's the advantage of this?

Reply:
answered offline.  this ensures our tokens follow the syntactic-model invariants of the roslyn tree system.